lib/maths/CTreeShapFeatureImportance.cc

```diff
@@ -29,19 +30,25 @@ void CTreeShapFeatureImportance::shap(core::CDataFrame& frame,
         maxDepthVec.push_back(maxDepth);
         m_SamplesPerNode.emplace_back(std::move(samplesPerNode));
     }
+    std::size_t maxDepthOverall =
+        *std::max_element(maxDepthVec.begin(), maxDepthVec.end());
 
-    auto result = frame.writeColumns(
-        m_NumberThreads, [&](const TRowItr& beginRows, const TRowItr& endRows) {
-            for (auto row = beginRows; row != endRows; ++row) {
-                auto encodedRow{encoder.encode(*row)};
-                for (std::size_t i = 0; i < m_Trees.size(); ++i) {
-                    SPath path(maxDepthVec[i] + 1);
-                    CTreeShapFeatureImportance::shapRecursive(
-                        m_Trees[i], m_SamplesPerNode[i], encoder, encodedRow,
-                        path, 0, 1.0, 1.0, -1, offset, row);
-                }
+    auto result = frame.writeColumns(m_NumberThreads, [&](const TRowItr& beginRows,
+                                                          const TRowItr& endRows) {
+        TElementVec pathVector;
+        TDoubleVec scaleVector;
+        // need a bit more memory than max depth
+        pathVector.reserve(((maxDepthOverall + 2) * (maxDepthOverall + 3)) / 2);
+        scaleVector.reserve(((maxDepthOverall + 2) * (maxDepthOverall + 3)) / 2);
```

Comment:
I also think it would be nice to note down the origin of `+ 2` and `+ 3`. I can see that why this is coming from the sum of an arithmetic progression in the worst case up to `maxDepthOverall`, but would be useful to explain this better. Also, at the same time you could explain the overall strategy of copying the "current path" to the end of each memory arena.

Reply:
I added a comment in ca3d8910